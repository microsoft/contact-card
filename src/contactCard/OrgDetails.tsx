import * as React from "react";
import { css, ActionButton, AnimationClassNames, Label, PersonaSize, Shimmer, ShimmerElementType } from "@fluentui/react";
import { Persona } from "../Persona";
import { PersonaShowMode, IPersonaProfile } from "../Types";


export function renderOrgHierarchy(profile: IPersonaProfile, allManagers: IPersonaProfile[] | undefined, directs: IPersonaProfile[] | undefined, onPersonaClick: (profile: IPersonaProfile) => void): React.ReactNode {
    return allManagers ?
        (
            <>
                <Label className="org-hierarchy-title">Organization</Label>
                <ul className={css("org-hierarchy", AnimationClassNames.slideLeftIn400)}>
                    {renderAllManagers(allManagers, onPersonaClick)}
                    <li className="person-current">
                        <Persona
                            id={profile.id}
                            displayName={profile.displayName}
                            showMode={PersonaShowMode.NameTitleDepartment}
                            size={PersonaSize.size48}
                        />
                    </li>
                    {renderDirects(directs, onPersonaClick)}
                </ul>
            </>
        ) :
        (
            <ul className={css("org-hierarchy", AnimationClassNames.slideLeftIn400)}>
                <li className="person"><Shimmer shimmerElements={[{ type: ShimmerElementType.circle, height: 48 }, { type: ShimmerElementType.gap, width: 12 }, { type: ShimmerElementType.line }]} width={"80%"} /></li>
                <li className="hierarchy-splitter" />
                <li className="person"><Shimmer shimmerElements={[{ type: ShimmerElementType.circle, height: 48 }, { type: ShimmerElementType.gap, width: 12 }, { type: ShimmerElementType.line }]} width={"80%"} /></li>
                <li className="hierarchy-splitter" />
                <li className="person"><Shimmer shimmerElements={[{ type: ShimmerElementType.circle, height: 48 }, { type: ShimmerElementType.gap, width: 12 }, { type: ShimmerElementType.line }]} width={"80%"} /></li>
            </ul>
        );
}


function renderAllManagers(allManagers: IPersonaProfile[], onPersonaClick: (profile: IPersonaProfile) => void): React.ReactNode[] {
    const res: React.ReactNode[] = [];

    for (let i = allManagers.length - 1; i >= 0; --i) {
        const mngr = allManagers[i];
        res.push(
            <li key={i}>
                <ActionButton className="person manager" onClick={() => onPersonaClick(mngr)}>
                    <Persona
                        id={mngr.id}
                        displayName={mngr.displayName}
                        showMode={PersonaShowMode.NameTitleDepartment}
                        size={PersonaSize.size48}
                    />
                </ActionButton>
            </li>
        );
        res.push(<li className="hierarchy-splitter" key={`spl-${i}`} />);
    }

    return res;
}


function renderDirects(directs: IPersonaProfile[] | undefined, onPersonaClick: (profile: IPersonaProfile) => void): React.ReactNode {
    if (!directs || !directs.length) {
        return <></>;
    }

    let key = 0;
    return (
        <>
            <li className="hierarchy-splitter" />
            <li className="directs">
                <Label className="directs-label">Direct reports ({directs.length})</Label>
                <ul>
                    {directs.map(direct => (
                        <li key={key++}>
                            <ActionButton className="person direct" onClick={() => onPersonaClick(direct)}>
                                <Persona
                                    id={direct.id}
                                    displayName={direct.displayName}
                                    showMode={PersonaShowMode.NameTitle}
                                    size={PersonaSize.size48}
                                />
                            </ActionButton>
                        </li>
                    ))}
                </ul>
            </li>
        </>
    );
}
