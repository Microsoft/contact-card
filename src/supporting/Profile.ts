import { IPersonaProfile } from "../Types";


export function buildProfileResponse(id: number) {
    return {
        id: `userId${id}`,
        displayName: `Display Name ${id}`,
        jobTitle: `Job Title ${id}`,
        department: `Department ${id}`,
        mail: `user${id}@contoso.com`,
        officeLocation: `1234-${id}`,
        city: "Redmond",
        businessPhones: [`555-123-456${id}`],
        imAddresses: [`im-user${id}@contoso.com`],
        companyName: "Contoso"
    };
}


export function buildProfile(id: number): IPersonaProfile {
    return {
        id: `userId${id}`,
        displayName: `Display Name ${id}`,
        jobTitle: `Job Title ${id}`,
        department: `Department ${id}`,
        email: `user${id}@contoso.com`,
        officeLocation: `1234-${id}`,
        city: "Redmond",
        businessPhone: `555-123-456${id}`,
        imAddress: `im-user${id}@contoso.com`,
        companyName: "Contoso"
    };
}
