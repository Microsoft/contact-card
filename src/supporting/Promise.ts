export async function delay(ms: number) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function waitNextTick() {
    return delay(0);
}
