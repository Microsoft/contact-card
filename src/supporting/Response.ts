import { buildProfileResponse } from "./Profile";

export function buildResponse(userId: number, status: number) {
    return new Response(JSON.stringify({
        responses: [
            {
                id: userId,
                body: buildProfileResponse(userId),
                status: status
            }
        ]
    }));
}
