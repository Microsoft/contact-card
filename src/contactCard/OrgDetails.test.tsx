import * as React from "react";
import * as Enzyme from "enzyme";
import { renderOrgHierarchy } from "./OrgDetails";
import { buildProfile } from "../supporting/Profile";


test("renders full OrgDetails", () => {
    const wrap = Enzyme.shallow(<div>{renderOrgHierarchy(buildProfile(1), [buildProfile(2), buildProfile(3)], [buildProfile(4), buildProfile(5)], jest.fn())}</div>);
    expect(wrap).toMatchSnapshot();
});


test("renders OrgDetails w/o managers", () => {
    const wrap = Enzyme.shallow(<div>{renderOrgHierarchy(buildProfile(1), undefined, undefined, jest.fn())}</div>);
    expect(wrap).toMatchSnapshot();
});


test("renders OrgDetails w/o directs", () => {
    const wrap = Enzyme.shallow(<div>{renderOrgHierarchy(buildProfile(1), [buildProfile(2), buildProfile(3)], undefined, jest.fn())}</div>);
    expect(wrap).toMatchSnapshot();
});


test("OrgDetails handle clicks", () => {
    const onPersonaClick = jest.fn();
    const wrap = Enzyme.shallow(<div>{renderOrgHierarchy(buildProfile(1), [buildProfile(2), buildProfile(3)], [buildProfile(4), buildProfile(5)], onPersonaClick)}</div>);

    wrap.find(".person.manager").first().simulate("click");
    expect(onPersonaClick).toBeCalledWith(buildProfile(3));

    wrap.find(".person.direct").first().simulate("click");
    expect(onPersonaClick).toBeCalledWith(buildProfile(4));
});
