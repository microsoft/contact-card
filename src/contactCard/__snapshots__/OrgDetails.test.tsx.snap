// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders OrgDetails w/o directs 1`] = `
<div>
  <StyledLabelBase
    className="org-hierarchy-title"
  >
    Organization
  </StyledLabelBase>
  <ul
    className="org-hierarchy css-40"
  >
    <li
      key="1"
    >
      <CustomizedActionButton
        className="person manager"
        onClick={[Function]}
      >
        <Persona
          displayName="Display Name 3"
          id="userId3"
          showMode={7}
          size={13}
        />
      </CustomizedActionButton>
    </li>
    <li
      className="hierarchy-splitter"
      key="spl-1"
    />
    <li
      key="0"
    >
      <CustomizedActionButton
        className="person manager"
        onClick={[Function]}
      >
        <Persona
          displayName="Display Name 2"
          id="userId2"
          showMode={7}
          size={13}
        />
      </CustomizedActionButton>
    </li>
    <li
      className="hierarchy-splitter"
      key="spl-0"
    />
    <li
      className="person-current"
    >
      <Persona
        displayName="Display Name 1"
        id="userId1"
        showMode={7}
        size={13}
      />
    </li>
  </ul>
</div>
`;

exports[`renders OrgDetails w/o managers 1`] = `
<div>
  <ul
    className="org-hierarchy css-40"
  >
    <li
      className="person"
    >
      <StyledShimmerBase
        shimmerElements={
          Array [
            Object {
              "height": 48,
              "type": 2,
            },
            Object {
              "type": 3,
              "width": 12,
            },
            Object {
              "type": 1,
            },
          ]
        }
        width="80%"
      />
    </li>
    <li
      className="hierarchy-splitter"
    />
    <li
      className="person"
    >
      <StyledShimmerBase
        shimmerElements={
          Array [
            Object {
              "height": 48,
              "type": 2,
            },
            Object {
              "type": 3,
              "width": 12,
            },
            Object {
              "type": 1,
            },
          ]
        }
        width="80%"
      />
    </li>
    <li
      className="hierarchy-splitter"
    />
    <li
      className="person"
    >
      <StyledShimmerBase
        shimmerElements={
          Array [
            Object {
              "height": 48,
              "type": 2,
            },
            Object {
              "type": 3,
              "width": 12,
            },
            Object {
              "type": 1,
            },
          ]
        }
        width="80%"
      />
    </li>
  </ul>
</div>
`;

exports[`renders full OrgDetails 1`] = `
<div>
  <StyledLabelBase
    className="org-hierarchy-title"
  >
    Organization
  </StyledLabelBase>
  <ul
    className="org-hierarchy css-40"
  >
    <li
      key="1"
    >
      <CustomizedActionButton
        className="person manager"
        onClick={[Function]}
      >
        <Persona
          displayName="Display Name 3"
          id="userId3"
          showMode={7}
          size={13}
        />
      </CustomizedActionButton>
    </li>
    <li
      className="hierarchy-splitter"
      key="spl-1"
    />
    <li
      key="0"
    >
      <CustomizedActionButton
        className="person manager"
        onClick={[Function]}
      >
        <Persona
          displayName="Display Name 2"
          id="userId2"
          showMode={7}
          size={13}
        />
      </CustomizedActionButton>
    </li>
    <li
      className="hierarchy-splitter"
      key="spl-0"
    />
    <li
      className="person-current"
    >
      <Persona
        displayName="Display Name 1"
        id="userId1"
        showMode={7}
        size={13}
      />
    </li>
    <li
      className="hierarchy-splitter"
    />
    <li
      className="directs"
    >
      <StyledLabelBase
        className="directs-label"
      >
        Direct reports (
        2
        )
      </StyledLabelBase>
      <ul>
        <li
          key="0"
        >
          <CustomizedActionButton
            className="person direct"
            onClick={[Function]}
          >
            <Persona
              displayName="Display Name 4"
              id="userId4"
              showMode={3}
              size={13}
            />
          </CustomizedActionButton>
        </li>
        <li
          key="1"
        >
          <CustomizedActionButton
            className="person direct"
            onClick={[Function]}
          >
            <Persona
              displayName="Display Name 5"
              id="userId5"
              showMode={3}
              size={13}
            />
          </CustomizedActionButton>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
